leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/endpoint/JavaCoapServerEndpoint.java

```diff
@@ -0,0 +1,233 @@
+package org.eclipse.leshan.transport.javacoap.endpoint;
+
+import java.net.URI;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.leshan.core.endpoint.Protocol;
+import org.eclipse.leshan.core.observation.Observation;
+import org.eclipse.leshan.core.observation.SingleObservation;
+import org.eclipse.leshan.core.request.DownlinkRequest;
+import org.eclipse.leshan.core.request.ObserveRequest;
+import org.eclipse.leshan.core.request.exception.SendFailedException;
+import org.eclipse.leshan.core.response.ErrorCallback;
+import org.eclipse.leshan.core.response.LwM2mResponse;
+import org.eclipse.leshan.core.response.ObserveResponse;
+import org.eclipse.leshan.core.response.ResponseCallback;
+import org.eclipse.leshan.server.endpoint.LwM2mServerEndpoint;
+import org.eclipse.leshan.server.endpoint.ServerEndpointToolbox;
+import org.eclipse.leshan.server.observation.LwM2mNotificationReceiver;
+import org.eclipse.leshan.server.profile.ClientProfile;
+import org.eclipse.leshan.server.registration.RegistrationStore;
+import org.eclipse.leshan.server.request.LowerLayerConfig;
+import org.eclipse.leshan.transport.javacoap.observation.ObservationUtil;
+
+import com.mbed.coap.client.CoapClient;
+import com.mbed.coap.client.CoapClientBuilder;
+import com.mbed.coap.client.ObservationConsumer;
+import com.mbed.coap.packet.CoapRequest;
+import com.mbed.coap.packet.CoapResponse;
+import com.mbed.coap.packet.Opaque;
+import com.mbed.coap.server.CoapServer;
+
+public class JavaCoapServerEndpoint implements LwM2mServerEndpoint {
+
+    private final URI endpointUri;
+    private final CoapServer coapServer;
+    private final ServerCoapMessageTranslator translator;
+    private final ServerEndpointToolbox toolbox;
+    private final LwM2mNotificationReceiver notificationReceiver;
+    private final RegistrationStore registrationStore;
+
+    public JavaCoapServerEndpoint(URI endpointUri, CoapServer coapServer, ServerCoapMessageTranslator translator,
+            ServerEndpointToolbox toolbox, LwM2mNotificationReceiver notificationReceiver,
+            RegistrationStore registrationStore) {
+        this.endpointUri = endpointUri;
+        this.coapServer = coapServer;
+        this.translator = translator;
+        this.toolbox = toolbox;
+        this.notificationReceiver = notificationReceiver;
+        this.registrationStore = registrationStore;
+
+    }
+
+    @Override
+    public Protocol getProtocol() {
+        return Protocol.COAP;
+    }
+
+    @Override
+    public URI getURI() {
+        return endpointUri;
+    }
+
+    @Override
+    public <T extends LwM2mResponse> T send(ClientProfile destination, DownlinkRequest<T> request,
+            LowerLayerConfig lowerLayerConfig, long timeoutInMs) throws InterruptedException {
```

Comment:
Is it really so, that each request can have different timeout?

Reply:
I see at least one use case : you don't want same timeout when you send "classic" request and when you send request which involved large payload (in request or response) using blockwise transfer. 

See : https://github.com/eclipse/leshan/wiki/Request-Timeout

I guess this should be easily implemented as most transport layer library should propose a way to cancel a request. 
(says the maintainer of the library which does not provide API to cancel LWM2M request :grin:)

Reply:
I see, maybe I could figure out a way to pass timeout on each request, which will overwrite default timeout.

Reply:
I don't know if this will work but my plan was do something like : 
```
public <T> CompletableFuture<T> timeoutAfter(long timeout, TimeUnit unit) {
    CompletableFuture<T> result = new CompletableFuture<T>();
    delayer.schedule(() -> result.completeExceptionally(new TimeoutException()), timeout, unit);
    return result;
}
```
OR to look how CompletableFuture.orTimeout is implemented. 

Following wording at https://github.com/eclipse/leshan/wiki/Request-Timeout , java-coap timeout is which kind of timeout ?

Reply:
I was talking about response timeout. (btw, very good explanation 👍)